### PR TITLE
perf(megatron): skip overwritten grad clears and isolate gradient reduce-scatter

### PR DESCRIPTION
**Pull request by suachong**
## Summary

- zero only the complement of gradient-buffer slices proven to be fully overwritten by Primus-Turbo beta=0 expert wgrad epilogues
- match ownership claims to exact parameter buffer offsets, addresses, dtypes, and element counts
- retain normal clearing for unowned parameters, alignment gaps, bucket padding, and every fallback path
- route gradient reduce-scatter through the dedicated zero-CTA RCCL group

## Why

Primus-Turbo can overwrite the complete `main_grad` view for the first MXFP4 expert wgrad write of an optimizer step. Clearing those same slices immediately before backward is redundant. This patch consumes Turbo's write-time ownership log and clears only the rest of Megatron's flat gradient buffer.

The gradient reduce-scatter is also isolated in the dedicated zero-CTA RCCL group so its SDMA-oriented path does not alter unrelated collectives.

## Correctness and safety

- iteration 0 and any run without ownership records clear the full buffer
- ownership is rebuilt from actual beta=0 writes every iteration
- multi-microbatch paths remain unowned and retain normal clearing
- skipped slices must match the buffer index map exactly
- alignment gaps and bucket-tail padding are always cleared
- the next iteration verifies every skipped slice was actually overwritten and raises on a missing producer
- optional poison mode fills skipped slices with NaN for validation

## Validation

- exercised with the GPT-OSS MXFP4 training path on MI355X
- verified production one-microbatch, forced multi-microbatch, and zero-token-expert behavior
- verified normal fallback when ownership is disabled or unavailable
- verified gradient-reduction group isolation and collective fallback behavior

## Dependency

Requires AMD-AGI/Primus-Turbo#534, which provides `primus_turbo.pytorch.core.grad_ownership` and records actual beta=0 overwrites.

### CODE REVIEW

**Comment by HuangWei-95** [comment]
need update the turbo commit in ci.yaml after the turbo's PR is merged

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import os
+from typing import Dict, FrozenSet, List, Tuple
+
+from primus.backends.megatron.patches.turbo.utils import is_primus_turbo_can_patch
+from primus.core.patches import PatchContext, get_args, register_patch
+from primus.core.utils.module_utils import log_rank_0
+
+Slice = Tuple[int, int]
+
+_DISABLED = os.environ.get("PRIMUS_TURBO_GRAD_OWNERSHIP", "1") != "1"
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py]: Fixed in 25fc56936. PRIMUS_TURBO_GRAD_OWNERSHIP now defaults to 0 and the module documentation states that setting it to 1 is required to opt in.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+_POISON = os.environ.get("PRIMUS_TURBO_GRAD_OWNERSHIP_POISON", "0") == "1"
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:69]: Fixed in 25fc56936. The correctness-sensitive skip is now explicit opt-in and remains disabled when the environment variable is absent.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+# Rebuilt every iteration by the wrapped ``zero_grad_buffer``; consumed by the
+# wrapped ``reset`` of every buffer belonging to that iteration.
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:72]: Fixed in 25fc56936. The gate now uses a safe default of 0, so this patch cannot activate silently.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:69]: Fixed in 25fc56936. The environment default and docstring now both describe explicit opt-in behavior.

File: primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    from megatron.core.distributed.param_and_grad_buffer import shard_buffer
+    from torch.distributed.distributed_c10d import _coalescing_manager
+
+    from primus.backends.megatron.core.distributed.rccl_sdma_param_gather import (
+        get_sdma_process_group,
+        is_direct_param_buffer,
+    )
+
+    @functools.wraps(original)
+    def start_grad_sync(self, force_all_reduce: bool = False):
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py:186]: Fixed in 25fc56936. Both custom-path exits now explicitly return None; the ineligible delegation still returns the original result.

File: primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        allocation_thread = threading.get_ident()
+        grad_data_allocated = False
+
+        def allocate_grad_data(*zeros_args, **zeros_kwargs):
+            nonlocal grad_data_allocated
+            if threading.get_ident() == allocation_thread and not grad_data_allocated:
+                grad_data_allocated = True
+                try:
+                    with torch.cuda.use_mem_pool(pool):
+                        return _REAL_TORCH_ZEROS(*zeros_args, **zeros_kwargs)
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py:141]: Fixed in 25fc56936. The inner parameter wrapper now uses its import-time real allocator for the internal param_data pool allocation, so the outer wrapper receives the constructor second zeros call for grad_data. Added a composed-wrapper regression proving both buffers allocate inside the pool and are marked direct.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        if start1 < end0:
+            raise RuntimeError(f"overlapping owned gradient slices ending {end0} and starting {start1}")
+    return found
+
+
+def _reset_complement(buffer) -> bool:
+    """Zero everything in ``buffer`` except the owned slices. False if it declined."""
+    slices = _owned_slices(buffer)
+    if not slices:
+        return False
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:181]: Covered by test_grad_buffer_ownership.py: empty and partial ownership, exact address/dtype/numel/contiguity matching, overlap rejection, complement and tail clearing, poison mode, step rotation, and skipped-but-unwritten failure are all exercised.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    original_zero_grad_buffer = DistributedDataParallel.zero_grad_buffer
+    original_reset = _ParamAndGradBuffer.reset
+
+    def zero_grad_buffer(self):
+        seen = _state["seen"]
+        if id(self) in seen or not seen:
+            _state["owned"] = grad_ownership.begin_step()
+            seen.clear()
+        seen.add(id(self))
+        return original_zero_grad_buffer(self)
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:265]: Fixed in 25fc56936. Ownership epochs rotated while curr_iteration is -1 are explicitly discarded. The warmup regression records a synthetic overwrite, discards it, and verifies that iteration 0 still receives an empty ownership set and therefore fully clears.

File: primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+            return None
+        assert (
+            self.grad_reduce_handle is None
+        ), "Should not have multiple communication calls outstanding at once"
+
+        if self.ddp_config.check_for_nan_in_grad or self.ddp_config.check_for_large_grads:
+            self.check_grads(
+                check_for_nan_or_inf=self.ddp_config.check_for_nan_in_grad,
+                check_for_large=self.ddp_config.check_for_large_grads,
+            )
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py:238]: Addressed in 66a453f17. The custom start_grad_sync now mirrors native Megatron by deduplicating and waiting on each parameter's _cudagraph_wgrad_ready_event before any gradient read, scaling, or reduce-scatter. Added an ordering regression proving the wait precedes the collective.

File: primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+        for bucket in self.buckets:
+            if bucket.gradient_scaling_factor != 1.0:
+                bucket.grad_data *= bucket.gradient_scaling_factor
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/parallelism/rccl_sdma_grad_reduce_scatter_patches.py]: Addressed in 66a453f17. The custom path now copies params_with_extra_main_grads from param.main_grad into main_grad_copy_in_grad_buffer under torch.no_grad(), before checks, scaling, and reduce-scatter, matching native Megatron. Added a regression that verifies copied data is scaled and visible to the collective.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    grad = buffer.grad_data
+    numel = grad.numel()
+    cursor = 0
+    for start, end, _key in slices:
+        if start > cursor:
+            grad[cursor:start].zero_()
+        cursor = end
+    if cursor < numel:
+        grad[cursor:numel].zero_()
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:191]: Addressed in 66a453f17. _reset_complement now explicitly zeroes every extra_main_grads tensor when it takes the optimized path, preserving the side effect of native _ParamAndGradBuffer.reset. Added a regression that confirms the owned slice remains untouched while the extra accumulation buffer is cleared.

**Comment by suachong** [comment]
Updated the PR Validation section to address the latest evidence/architecture reviews with: (1) the exact 8×MI355X workload and pinned baseline/candidate commits, (2) median/mean/p95/HBM/NaN/skipped A/B results, including the repeated committed-tree run, (3) campaign, baseline-profile, and candidate-profile links, (4) copy-pasteable campaign and 8-rank reduce-scatter probe commands, (5) production-shape CTA-vs-CE latency and numerical-correctness results, and (6) the rationale for retaining the explicit-opt-in always-async RCCL workaround until an authoritative fixed runtime version is identified. The code review’s three native-Megatron parity findings are fixed in 66a453f17 with regressions.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+def _is_enabled(ctx: PatchContext) -> bool:
+    if _DISABLED:
+        return False
+    args = get_args(ctx)
+
+    # Ownership recording happens in Python after the beta=0 custom op and is
+    # therefore not replayed by CUDA graphs. Keep full framework clearing for
+    # any configured graph mode.
+    if bool(getattr(args, "enable_cuda_graph", False)) or bool(getattr(args, "external_cuda_graph", False)):
+        return False
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py]: Addressed in 0a60809c1. The ownership consumer now checks graph configuration on the direct Megatron args and nested model_cfg/model_config/model objects. It disables the optimization for top-level graph flags and any non-disabled cuda_graph_impl, including local and transformer_engine. Added regressions for both implementations across every supported config location, plus disabled-value coverage.

**Comment by suachong** [comment]
On the latest test-coverage report: this PR does include committed regressions in `tests/unit_tests/backends/test_rccl_sdma_grad_reduce_scatter.py` and `tests/unit_tests/backends/megatron/test_grad_buffer_ownership.py` (routing/fallback, scaling, event ordering, promoted-main-grad staging, allocation composition, exact ownership matching, poison/fail-closed behavior, warmup and graph gates, and extra-gradient clearing). The hardware collective itself was validated with the linked 8-rank MI355X production-shape probe and 120-step training smoke; the PR description now contains the exact commands and absolute A/B results. GPT-OSS communicates BF16 gradients here—FP4 describes weight storage, not the reduce-scatter buffer dtype—so an FP4 gradient-buffer case would not represent this path. A mandatory multi-GPU CI job would require repository GPU-runner policy/infrastructure beyond this PR; the target-hardware run links are provided for reproducibility and review.

File: primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    original_reset = _ParamAndGradBuffer.reset
+    args = get_args(ctx)
+
+    def zero_grad_buffer(self):
+        seen = _state["seen"]
+        if id(self) in seen or not seen:
+            _state["owned"] = _rotate_ownership(
+                grad_ownership,
+                discard=getattr(args, "curr_iteration", None) == -1,
+            )
```
[the remark this answers is not in the document]
Comment (suachong) [reply primus/backends/megatron/patches/turbo/grad_buffer_ownership_patches.py:262]: Addressed in Primus 2fe1f05f0 and Primus-Turbo 445b6664. The consumer now discards previous ownership (and performs the full clear) unless the current Megatron schedule reports exactly one microbatch, failing closed if the schedule is unavailable. It also validates each bucket group immediately before start_grad_sync: every slice skipped during reset must have a matching beta=0 overwrite in the current iteration before that bucket can be reduced. This catches a disappearing or beta=1 producer before the optimizer can consume a stale reduced gradient; the next-iteration check remains as a secondary safety net. Added focused regressions for multi-microbatch/unknown schedules, stale current-step slices, overwritten slices, and unrelated later buckets.

File: tests/unit_tests/backends/megatron/test_grad_buffer_ownership.py [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+built inside ``patch_grad_buffer_ownership()``, which patches real Megatron
+classes, so exercising them needs an actual (or end-to-end) Megatron
+environment rather than a unit test.
+"""
+
+from types import SimpleNamespace
+
+import pytest
+import torch
+from primus_turbo.pytorch.core import grad_ownership
```
[the remark this answers is not in the document]
Comment (suachong) [reply tests/unit_tests/backends/megatron/test_grad_buffer_ownership.py]: Fixed in e13375a5e. The test now loads primus_turbo.pytorch.core.grad_ownership through pytest.importorskip with an explicit optional-dependency reason, so Primus test collection succeeds and this Turbo-dependent module is skipped when Turbo is intentionally absent.
